### PR TITLE
Add duckdb support

### PR DESCRIPTION
**Pull request by sunank200**
# Description
## What is the current behavior?

Currently Duckdb is not supported with astro-sdk-python

closes: #1130
closes: #1138

## What is the new behavior?

- Add duckdb support 
- Add the tests
- Update the documentation
- Add example DAG

## Does this introduce a breaking change?
No

### Checklist
- [x] Created tests which fail without the change (if possible)
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: python-sdk/docs/astro/sql/operators/load_file.rst [hunk 1]
```diff
@@ -221,6 +221,17 @@ Loading to MS SQL
 .. note::
    We do not support loading Unicode data to SQL Server due to limitations on the underlying ``pymssql`` library
 
+Loading to Duckdb
+~~~~~~~~~~~~~~~~~
+
+``load_file`` can load data to duckdb. If the database file is not specified in connection, it will be created
+(the file extension may be .db, .duckdb, or anything else). The special value :memory: (the default) can be used to
```
Comment (tatiana) [inline python-sdk/docs/astro/sql/operators/load_file.rst:228]: What determines the file extension?

File: python-sdk/src/astro/databases/duckdb.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        self.table = table
+        self.load_options = load_options
+
+    @property
+    def sql_type(self) -> str:
+        return "duckdb"
+
+    @cached_property
+    def hook(self) -> DuckDBHook:
+        """Retrieve Airflow hook to interface with the Sqlite database."""
```
Comment (tatiana) [inline python-sdk/src/astro/databases/duckdb.py:36]:
```suggestion
        """Retrieve Airflow hook to interface with the DuckDB database."""
```

File: python-sdk/tests_integration/databases/test_all_databases.py [hunk 3]
```diff
@@ -118,9 +123,10 @@ def test_export_table_to_pandas_dataframe_mssql(
         {"database": Database.SQLITE},
         {"database": Database.DELTA},
         {"database": Database.MSSQL},
+        {"database": Database.DUCKDB},
     ],
     indirect=True,
-    ids=["bigquery", "postgres", "redshift", "snowflake", "sqlite", "delta", "mssql"],
+    ids=["bigquery", "postgres", "redshift", "snowflake", "sqlite", "delta", "mssql", "duckdb"],
```
Comment (tatiana) [inline python-sdk/tests_integration/databases/test_all_databases.py:129]: By looking at how many places you had to make this type of change, I was wondering if we could, by any chance, try to compute things using `SUPPORTED_DATABASES`:
https://github.com/astronomer/astro-sdk/blob/main/python-sdk/src/astro/constants.py#L70

So we don't need the list hard-coded in so many tests. If we managed to get this to work, it would be much easier to add future databases, since we wouldn't need to update as many tests

File: python-sdk/docs/astro/sql/operators/load_file.rst (hunk 1 above)
Comment (tatiana) [reply python-sdk/docs/astro/sql/operators/load_file.rst:228]: It may be worth to highlight that the value `memory` should be set in the connection, so users don't expect to set it in the `load_file` task itself.

File: python-sdk/docs/astro/sql/operators/load_file.rst [hunk 4]
```diff
@@ -221,6 +221,17 @@ Loading to MS SQL
...
+create an in-memory database. Note that for an in-memory database no data is persisted to disk (i.e. all data is lost when you exit the Python process).
```
Comment (jlaneve) [inline python-sdk/docs/astro/sql/operators/load_file.rst:229]: what's the behavior like if you load a file into duckdb and the next task gets executed on a different worker? somewhat separately, what's the behavior like for `load_file` with sqlite? I'm a bit concerned about the case where things get executed on separate workers if you're storing data in tmp storage or in memory

File: python-sdk/tests_integration/databases/test_all_databases.py (hunk 3 above)
Comment (utkarsharma2) [reply python-sdk/tests_integration/databases/test_all_databases.py:129]: Maybe this can be helpful as well - https://hackebrot.github.io/pytest-tricks/param_id_func/

File: python-sdk/pyproject.toml [hunk 5]
```diff
@@ -94,6 +94,11 @@ mssql = [
     "apache-airflow-providers-microsoft-mssql>=3.2",
 ]
 
+duckdb = [
+    "airflow-provider-duckdb",
+    "duckdb-engine"
```
Comment (jlaneve) [inline python-sdk/pyproject.toml:99]: should duckdb-engine be a dependency of the SDK or of the duckdb airflow provider?

File: python-sdk/src/astro/databases/duckdb.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    def schema_exists(self, schema: str) -> bool:  # skipcq PYL-W0613,PYL-R0201
+        """
+        Check if a schema exists. We return false for duckdb since duckdb does not have schemas
+        """
+        return False
+
+    @staticmethod
+    def get_merge_initialization_query(parameters: tuple) -> str:
+        """
+        Handles database-specific logic to handle index for Sqlite.
```
Comment (jlaneve) [inline python-sdk/src/astro/databases/duckdb.py:78]:
```suggestion
        Handles database-specific logic to handle index for DuckDB.
```

**Review by jlaneve — COMMENTED** [review]
this looks great! can we add an example dag that looks something like:
1. python task that returns a dataframe
2. aql run_raw_sql/transform that runs a query on the dataframe
3. python task that reads the results back in as a df

File: python-sdk/pyproject.toml (hunk 5 above)
Comment (feluelle) [reply python-sdk/pyproject.toml:99]: It shouldn't unless we need to patch the provider temporarily. Ideally, we should only depend on the provider.

File: python-sdk/tests_integration/databases/test_all_databases.py (hunk 3 above)
Comment (feluelle) [reply python-sdk/tests_integration/databases/test_all_databases.py:129]: I agree. We should have a marker that makes the test case run for all databases, without us having to redefine databases or ids.

Comment (sunank200) [reply python-sdk/tests_integration/databases/test_all_databases.py:129]: Should this change be part of this PR? I think that can be a separate issue. WDYT?

File: python-sdk/pyproject.toml (hunk 5 above)
Comment (sunank200) [reply python-sdk/pyproject.toml:99]: Yes duckdb-engine should have this dependency. It's just a temporary solution. I have already create a PR for this: https://github.com/astronomer/airflow-provider-duckdb/pull/2

Comment (sunank200) [reply python-sdk/pyproject.toml:99]: This can be removed once this change is released on providers side

File: python-sdk/docs/astro/sql/operators/load_file.rst (hunk 4 above)
Comment (sunank200) [reply python-sdk/docs/astro/sql/operators/load_file.rst:229]: Users have the option to set a host if they want to use shared database file locations across workers. This is similar to the sqlite database.

With `:memory:` no data is persisted to disk (i.e. all data is lost when you exit the Python process).

File: python-sdk/example_dags/example_load_file.py [hunk 7]
```diff
@@ -353,4 +354,13 @@
     )
     # [END load_file_example_26]
 
+    # [START load_file_example_27]
+    aql.load_file(
+        input_file=File("s3://tmp9/homes_main.csv", conn_id="aws_conn"),
```
[the remark this answers is not in the document]
Comment (sunank200) [reply python-sdk/example_dags/example_load_file.py]: Done. https://github.com/astronomer/astro-sdk/pull/1695/commits/bb4a8288e7224a496153bdcd93cbef7655034d02

File: python-sdk/docs/astro/sql/operators/load_file.rst (hunk 1 above)
Comment (sunank200) [reply python-sdk/docs/astro/sql/operators/load_file.rst:228]: Screenshot 2023-02-02 at 10 03 44 PM
In connections if the file is kept empty it will use `memory`.

File: python-sdk/pyproject.toml (hunk 5 above)
Comment (sunank200) [reply python-sdk/pyproject.toml]: removed now: https://github.com/astronomer/astro-sdk/pull/1695/commits/a230d7316d24af6a98ed444557a21c4da4ae4609

**Comment by sunank200** [comment]
> this looks great! can we add an example dag that looks something like:
> 
> 1. python task that returns a dataframe
> 2. aql run_raw_sql/transform that runs a query on the dataframe
> 3. python task that reads the results back in as a df

@juliana-astro I have added an example. Please check

File: python-sdk/docs/astro/sql/operators/load_file.rst (hunk 1 above)
Comment (sunank200) [reply python-sdk/docs/astro/sql/operators/load_file.rst:228]: @tatiana these are the supported extensions: https://duckdb.org/docs/extensions/overview.html

**Review by pankajastro — COMMENTED** [review]
Please update https://github.com/astronomer/astro-sdk/blob/main/python-sdk/README.md#supported-technologies

File: python-sdk/src/astro/databases/duckdb.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        """
+        table.conn_id = table.conn_id or self.conn_id
+        return table
+
+    def create_schema_if_needed(self, schema: str | None) -> None:
+        """
+        Since Duckdb does not have schemas, we do not need to set a schema here.
+        """
+
+    def schema_exists(self, schema: str) -> bool:  # skipcq PYL-W0613,PYL-R0201
```
Comment (pankajastro) [inline python-sdk/src/astro/databases/duckdb.py:70]: can we decorate this method with `@staticmethod ` ?

**Comment by sunank200** [comment]
> Update readme

Added @pankajastro https://github.com/astronomer/astro-sdk/pull/1695/commits/8042faa66980648de3438f33e26f1a3a450c9c2a

File: python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    filtered_dataframe = filter_data(
+        adoption_center_data,
+    )
+
+    aggregated_dataframe = aggregate_data(
+        filtered_dataframe,
+        output_table=Table(conn_id=DUCKDB_CONN_ID),
+    )
+
+    s3_bucket = os.getenv("GCS_BUCKET", "s3://astro-sdk")
```
Comment (pankajastro) [inline python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py:58]: why `GCS_BUCKET` and path is s3 typo?

File: python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    return new_df
+
+
+@dag(
+    start_date=datetime(2021, 1, 1),
+    max_active_runs=1,
+    schedule_interval=None,
+    default_args={
+        "email_on_failure": False,
+        "retries": 0,
```
Comment (pankajastro) [inline python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py:37]: if retry=0 then does retry_delay make sense?

File: python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py (hunk 9 above)
Comment (pankajastro) [reply python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py:58]: what about moving this on top level?

File: python-sdk/src/astro/databases/duckdb.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+        Check if a schema exists. We return false for duckdb since duckdb does not have schemas
+        """
+        return False
+
+    @staticmethod
+    def get_merge_initialization_query(parameters: tuple) -> str:
+        """
+        Handles database-specific logic to handle index for DuckDB.
+        """
+        return "CREATE UNIQUE INDEX merge_index ON {{table}}(%s)" % ",".join(parameters)  # skipcq PYL-C0209
```
Comment (pankajastro) [inline python-sdk/src/astro/databases/duckdb.py:81]: why `skipcq PYL-C0209` can we use f-string here?

File: python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py (hunk 9 above)
Comment (sunank200) [reply python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py]: Changed: https://github.com/astronomer/astro-sdk/pull/1695/commits/a4245e67ffc51cf5f1c3a0616dedbb960fa9ab2e

File: python-sdk/src/astro/databases/duckdb.py (hunk 8 above)
Comment (sunank200) [reply python-sdk/src/astro/databases/duckdb.py]: Removed this completely.

File: python-sdk/src/astro/databases/duckdb.py (hunk 11 above)
Comment (sunank200) [reply python-sdk/src/astro/databases/duckdb.py]: Removed this completely

File: python-sdk/docs/astro/sql/operators/load_file.rst [hunk 12]
```diff
@@ -221,6 +221,17 @@ Loading to MS SQL
 .. note::
    We do not support loading Unicode data to SQL Server due to limitations on the underlying ``pymssql`` library
 
+Loading to Duckdb
+~~~~~~~~~~~~~~~~~
+
+``load_file`` can load data to duckdb. If the database file is not specified in connection, it will be created
```
Comment (tatiana) [inline python-sdk/docs/astro/sql/operators/load_file.rst:227]:
```suggestion
``load_file`` can load data to duckdb. If the database file is not specified in the connection, it will assume the user is using an in-memory duckdb. For more information about the supported files, check the [official documentation](https://duckdb.org/docs/extensions/overview.html).
```

File: python-sdk/docs/astro/sql/operators/load_file.rst (hunk 1 above)
Comment (tatiana) [reply python-sdk/docs/astro/sql/operators/load_file.rst:228]: @sunank200 thanks, based on what you wrote - we may want to rephrase the paragraph - please, take a look at the suggestion :)

File: python-sdk/tests_integration/databases/test_all_databases.py (hunk 3 above)
Comment (tatiana) [reply python-sdk/tests_integration/databases/test_all_databases.py:129]: Happy to have it logged as a separate ticket! Please, could you add the link here, @sunank200 , and help us prioritise it?

File: python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    new_df = df.pivot_table(index="DATE", values="NAME", columns=["TYPE"], aggfunc="count").reset_index()
+    new_df.columns = new_df.columns.str.lower()
+    return new_df
+
+
+@dag(
+    start_date=datetime(2021, 1, 1),
+    max_active_runs=1,
+    schedule_interval=None,
+    default_args={
```
Comment (pankajastro) [inline python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py:36]: `retry_delay` 5min looks lots of time. I rather like to remove `default_args`.  but it completely your call

**Review by pankajastro — APPROVED** [review]
LGTM

File: python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+@aql.dataframe()
+def aggregate_data(df: pd.DataFrame):
+    new_df = df.pivot_table(index="DATE", values="NAME", columns=["TYPE"], aggfunc="count").reset_index()
+    new_df.columns = new_df.columns.str.lower()
+    return new_df
+
+
+@dag(
+    start_date=datetime(2021, 1, 1),
+    max_active_runs=1,
```
Comment (pankajastro) [inline python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py:34]: why we need this?

File: python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py (hunk 13 above)
Comment (jlaneve) [reply python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py:36]: I agree - we should make this as minimal as possible to highlight just what we're trying to show (duckdb transformations) and remove any non-required fields (default_args, max_active_runs)

File: python-sdk/README.md [hunk 15]
```diff
@@ -106,6 +106,7 @@ pip install astro-sdk-python[amazon,google,snowflake,postgres]
 | Snowflake        |
 | SQLite           |
 | Amazon Redshift  |
+| Duckdb           |
```
Comment (jlaneve) [inline python-sdk/README.md:109]: nit:
```suggestion
| DuckDB           |
```

File: python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py (hunk 14 above)
Comment (sunank200) [reply python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py]: done https://github.com/astronomer/astro-sdk/pull/1695/commits/89a942d91af5885e387932194f289198b973506c

File: python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py (hunk 13 above)
Comment (sunank200) [reply python-sdk/example_dags/example_duckdb_load_transform_dataframe_and_save.py]: done https://github.com/astronomer/astro-sdk/pull/1695/commits/89a942d91af5885e387932194f289198b973506c

File: python-sdk/docs/astro/sql/operators/load_file.rst (hunk 12 above)
Comment (sunank200) [reply python-sdk/docs/astro/sql/operators/load_file.rst]: https://github.com/astronomer/astro-sdk/pull/1695/commits/e09109bc9c1daa83b99b36608ed066ef95a2e3e4

File: python-sdk/docs/astro/sql/operators/load_file.rst (hunk 1 above)
Comment (sunank200) [reply python-sdk/docs/astro/sql/operators/load_file.rst]: Done https://github.com/astronomer/astro-sdk/pull/1695/commits/e09109bc9c1daa83b99b36608ed066ef95a2e3e4

File: python-sdk/tests_integration/databases/test_all_databases.py (hunk 3 above)
Comment (sunank200) [reply python-sdk/tests_integration/databases/test_all_databases.py:129]: created the ticket here: https://github.com/astronomer/astro-sdk/issues/1711

File: python-sdk/tests_integration/sql/operators/test_merge.py [hunk 16]
```diff
@@ -452,8 +475,9 @@ def get_result_sql_single(conn_id):
         (base_database_class, "gcp_conn"),
         (base_database_class, "snowflake_conn"),
         (delta_database_class, "databricks_conn"),
+        (base_database_class, "duckdb_conn"),
     ],
-    ids=["sqlite", "bigquery", "snowflake", "databricks"],
+    ids=["sqlite", "bigquery", "snowflake", "databricks", "duckdb"],
 )
```
Comment (utkarsharma2) [inline python-sdk/tests_integration/sql/operators/test_merge.py:481]: Can we extend duck db for these tests? - test_merge_with_the_same_schema, run_merge, test_merge_with_different_schemas

Comment (sunank200) [reply python-sdk/tests_integration/sql/operators/test_merge.py:481]: Sure.

Comment (sunank200) [reply python-sdk/tests_integration/sql/operators/test_merge.py:485]: @utkarsharma2 added, please check: https://github.com/astronomer/astro-sdk/pull/1695/commits/d8bbd2c2314eaf5bb23df2ff83a7fbf9756be4bf

**Comment by sunank200** [comment]
@utkarsharma2 removing one test - `test_merge` and details about this are added on the issue https://github.com/astronomer/astro-sdk/issues/1713.

TL;DR; Current release of DuckDB has a bug https://github.com/duckdb/duckdb/issues/6087 which will be fixed in the new release.

**Review by tatiana — APPROVED** [review]

File: python-sdk/tests_integration/sql/operators/test_merge.py (hunk 16 above)
Comment (utkarsharma2) [reply python-sdk/tests_integration/sql/operators/test_merge.py:483]: Can we mention a comment in the code next to test cases that won't work for duck DB?

**Review by utkarsharma2 — APPROVED** [review]
